tests/capture/test_base_interpreter.py

```diff
@@ -0,0 +1,439 @@
+            def g(i):
+                qml.adjoint(qml.Z(i))
+                return i + 1
+
+            g(0)
+
+        jaxpr = jax.make_jaxpr(f)(3)
+        assert jaxpr.eqns[0].primitive == while_loop_prim
+
+        inner_jaxpr = jaxpr.eqns[0].params["jaxpr_body_fn"]
+        assert len(inner_jaxpr.eqns) == 2
+        assert inner_jaxpr.eqns[0].primitive == qml.Z._primitive  # no adjoint of x
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 3)
+
+        qml.assert_equal(q.queue[0], qml.Z(0))
+        qml.assert_equal(q.queue[1], qml.Z(1))
+        qml.assert_equal(q.queue[2], qml.Z(2))
+        assert len(q) == 3
+
+    def test_qnode(self):
+        """Test transforming qnodes."""
+
+        class AddNoise(PlxprInterpreter):
+
+            def interpret_operation(self, op):
+                new_op = op._unflatten(*op._flatten())
+                _ = [qml.RX(0.1, w) for w in op.wires]
+                return new_op
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @AddNoise()
+        @qml.qnode(dev, diff_method="adjoint", grad_on_execution=False)
+        def f():
+            qml.I(0)
+            qml.I(0)
+            return qml.probs(wires=0)
+
+        jaxpr = jax.make_jaxpr(f)()
+        assert jaxpr.eqns[0].primitive == qnode_prim
+        inner_jaxpr = jaxpr.eqns[0].params["qfunc_jaxpr"]
+
+        assert len(inner_jaxpr.eqns) == 5
+        assert inner_jaxpr.eqns[0].primitive == qml.I._primitive
+        assert inner_jaxpr.eqns[2].primitive == qml.I._primitive
+        assert inner_jaxpr.eqns[1].primitive == qml.RX._primitive
+        assert inner_jaxpr.eqns[3].primitive == qml.RX._primitive
+
+        assert jaxpr.eqns[0].params["qnode_kwargs"]["diff_method"] == "adjoint"
+        assert jaxpr.eqns[0].params["qnode_kwargs"]["grad_on_execution"] is False
+        assert jaxpr.eqns[0].params["device"] == dev
+
+        res1 = f()
+        # end up performing two rx gates with phase of 0.1 each on wire 0
+        expected = jax.numpy.array([jax.numpy.cos(0.2 / 2) ** 2, jax.numpy.sin(0.2 / 2) ** 2])
+        assert qml.math.allclose(res1, expected)
+        res2 = jax.core.eval_jaxpr(jaxpr.jaxpr, [])
+        assert qml.math.allclose(res2, expected)
+
+    @pytest.mark.parametrize("grad_f", (qml.grad, qml.jacobian))
+    def test_grad_and_jac(self, grad_f):
+        """Test interpreters can handle grad and jacobian HOP's."""
+
+        class DoubleAngle(PlxprInterpreter):
+
+            def interpret_operation(self, op):
+                leaves, struct = jax.tree_util.tree_flatten(op)
+                return jax.tree_util.tree_unflatten(struct, [2 * l for l in leaves])
+
+        @DoubleAngle()
+        def f(x):
+            @qml.qnode(qml.device("default.qubit", wires=2))
+            def circuit(y):
+                qml.RX(y, 0)
+                return qml.expval(qml.Z(0))
+
+            return grad_f(circuit)(x)
```

Comment:
Oh, this is a very interesting example!
In code, we have an inner function that uses `grad/jacobian`, and an outer function transforming this.
So a naive translation into math would be $\text{DoubleAngle}(\partial_x (f)) (x)$. However, we end up doubling the angle _first_ and differentiating afterwards, so $\partial_x (\text{DoubleAngle}(f))(x)$, right? 
I would not have expected this without knowing the internals of an interpreter. :thinking: 
If one wanted to have the interpreter outside the differentiation, I suppose one would have to register all primitives that do not commute with the interpreter's effect :thinking:

Reply:
I guess for me doubling the angle after differentiating just doesn't make sense. 

I can switch back to testing the `SimplifyInterpreter` and just testing the contents of the jaxpr are different.  That might be a better way of testing tthis.

Reply:
>I guess for me doubling the angle after differentiating just doesn't make sense.

ah, sure, I don't care too much about the concrete example per se, it's just an example where a HOP and an interpreter do not commute, and the code ordering and execution ordering differ. Certainly that will be a surprise one day :eyes:

Reply:
Looking at this again, I still find it confusing.
Can we move the `DoubleAngle` decorator into `f` and apply it to the QNode directly? The result will be the same, but notational order and evaluation order will match.

Reply:
I can instead use `SimplifyInterpreter` and do a f-> jaxpr evaluation and check the contents of the jaxpr.